### PR TITLE
 RPC design refactoring

### PR DESCRIPTION
**Pull request by LuxoftAKutsan**
Technical task : https://github.com/smartdevicelink/sdl_core/issues/2189

This PR change architecture document and describes the new approach of RPC plugin system in SDL.

SDL commands were part of the application manager component.  
New solution is to extract plugin with following responsibilities : 
 - Create certain Commands
 - Process certain commands logic
All SDL commands may be split into plugins. 
Architecture approach shows the high-level solution of keeping encapsulation between plugins.

Added additional entity: Plugin manager, that loads plugins from the filesystem during SDL startup.
From ApplicationManager extracted components: RPCService, RPCHandler. 
So ApplicationManager is not responsible for handling and sending RPCs anymore, t store only applications/system related data. 
Plugin may also contain some classes that related only to commands that it includes.
There is no common approach to get such internal data from a plugin and it is not a part of this refactoring. 
This refactoring main goal is separate plugins data and to avoid undesired access. 

Other parts of delivery :  
- SDL core: https://github.com/smartdevicelink/sdl_core/pull/2196 
- CI activities: http://opensdl-jenkins.luxoft.com:8080/view/Rc_refactoring/
- Architecture changes: https://github.com/smartdevicelink/sdl_core_documentation/pull/10

### CODE REVIEW

File: docs/Software Architecture Document/Components View/index.md [hunk 1]
```diff
@@ -102,15 +101,32 @@ The view is represented by module and subsystem diagrams that show the system's
...
+    - Requires ***ProtocolHandler** and **ProtocolObserver (Protocol Handler)***
+    - Requires ***HMIMessageObserver** and **HMIMessageSender (HMI Message Handler)***
+  - *Interfaces* 
+    - Provides ***RPCService*** interface
+  - *Behavior*
+    - The component implements RPC transfering functionality
+  - *Constraints*
+    - *N/A* 
+
 #### Commands
```
Comment (yang1070) [inline docs/Software Architecture Document/Components View/index.md]: where does the Commands component go to in the diagram? Seems it is no longer there.

File: docs/Software Architecture Document/Components View/index.md [hunk 2]
```diff
@@ -163,6 +179,18 @@ The view is represented by module and subsystem diagrams that show the system's
...
+#### Plugin
+  - *Responsibility*
+    - Process Mobile/HMI RPC
+  - *Relations*
+    - Composed by ***Plugin Manager*** 
+    - Uses ***Commands***
+  - *Interfaces*
+    - Provides ***Plugin*** interface 
+  - *Behavior*
+    - Handle pecific RPC and process it's internal logic
```
Comment (yang1070) [inline docs/Software Architecture Document/Components View/index.md]: typo pecific

File: docs/Software Architecture Document/Components View/index.md [hunk 3]
```diff
@@ -214,12 +242,12 @@ The view is represented by module and subsystem diagrams that show the system's
...
+    - Handles ***Application Manager*** by ***Application Manager*** interface
   - *Interfaces*
-    - Provides ***Plugin Manager*** interface   
+    - Provides ***Plugin*** interface   
   - *Behavior*
-    - Receives data from CoreService
+    - Receives data from ApplicationManager
     - Parses data
     - Creates commands.
     - Handles incoming HMI notifications
```
Comment (yang1070) [inline docs/Software Architecture Document/Components View/index.md]: any updates to sdl_core_documentation/docs/Software Architecture Document/Data View/?
  as PluginManger/Plugins/RPC service comes into the picture.

Comment (yang1070) [inline docs/Software Architecture Document/Components View/index.md]: also any updates to 
 section 4.8.2 in https://github.com/smartdevicelink/sdl_core_documentation/blob/develop/docs/Software%20Architecture%20Document/Development%20View/index.md ?

Comment (yang1070) [inline docs/Software Architecture Document/Components View/index.md]: aslo 
sdl_core_documentation/docs/Software Architecture Document/View-to-View Relations/index.md

**Comment by LitvinenkoIra** [comment]
@yang1070 please look at the latest updates 39b93f1

File: docs/Software Architecture Document/View-to-View Relations/index.md [hunk 4]
```diff
@@ -37,6 +39,7 @@ The following table is a mapping between the elements in the Data view and the D
 | ***SmartObject***     | X                     |                    |                     |              |
 | ***Mobile Command***  | X                     |                    |                     |              |
 | ***HMI Command***     | X                     |                    |                     |              |
+| ***RPS Service***     | X                     |                    |                     |              |
```
Comment (yang1070) [inline docs/Software Architecture Document/View-to-View Relations/index.md]: what is RPS?  is it `RPC Service`?

File: docs/Software Architecture Document/Components View/index.md [hunk 5]
```diff
@@ -13,135 +13,150 @@ The view is represented by module and subsystem diagrams that show the system's
...
+    - Requires ***ProtocolHandler** and **ProtocolObserver (Protocol Handler)***
+    - Requires ***HMIMessageObserver** and **HMIMessageSender (HMI Message Handler)***
+  - *Interfaces*
+    - Provides ***RPCService*** interface
+  - *Behavior*
+    - The component implements RPC transfering functionality
+  - *Constraints*
+    - *N/A*
 
 #### Commands
```
Comment (yang1070) [inline docs/Software Architecture Document/Components View/index.md]: do we still has a `Commands` component? I thought it is only an interface now

File: docs/Software Architecture Document/Components View/index.md [hunk 6]
```diff
@@ -13,135 +13,150 @@ The view is represented by module and subsystem diagrams that show the system's
...
+    - ***HMIMessageHandler*** interface for accumulating ***HMIMessageObserver, HMIMessageSender*** and ***HMIMessageAdapter***
   - *Behavior*
-    - Transferring RPC Messages between business-layer and configured transport. 
-  - *Constraints* 
+    - Transferring RPC Messages between business-layer and configured transport.
+  - *Constraints*
     - Processes messages from a single instance of HMI only.
     - HMI-transport need to be statically configurable with build flags.
 
 ### Business layer components:
```
Comment (yang1070) [inline docs/Software Architecture Document/Components View/index.md]: in the diagram, it is called `application` layer, not `business` layer, and there are 10 components in the diagram, but 12 components in this list. Something mismatch?

**Review by yang1070 — CHANGES_REQUESTED** [review]
please update

File: docs/Software Architecture Document/View-to-View Relations/index.md (hunk 4 above)
Comment (LuxoftAKutsan) [reply docs/Software Architecture Document/View-to-View Relations/index.md]: @yang1070 fixed 3f6450a

File: docs/Software Architecture Document/Components View/index.md (hunk 5 above)
Comment (LuxoftAKutsan) [reply docs/Software Architecture Document/Components View/index.md]: @yang1070 fixed 3f6450a

File: docs/Software Architecture Document/Components View/index.md (hunk 6 above)
Comment (LuxoftAKutsan) [reply docs/Software Architecture Document/Components View/index.md]: @yang1070 fixed 3f6450a

**Review by yang1070 — APPROVED** [review]
Reviewed during online meeting. Looks good to me.

**Comment by yang1070** [comment]
@theresalech  Ford has reviewed the changes and approves the PR.

**Review by Jack-Byrne — APPROVED** [review]

**Pull request merged**